cmd/clusterctl/pkg/client/cluster/move.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// objectReference defines a reference to a Kubernetes object that is visited during the discovery phase for the move operation.
+type objectReference struct {
+	APIVersion string
+	Kind       string
+	Namespace  string
+	Name       string
+	UID        types.UID
+}
+
+func (s objectReference) String() string {
+	return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
+}
+
+// node defines a node in the Kubernetes object graph that is visited during the discovery phase for the move operation.
+type node struct {
+	identity objectReference
+
+	// dependents contains the list of nodes that are owned by the current node.
+	// Nb. This is the reverse of metadata.OwnerRef.
+	dependents map[*node]struct{}
```

Comment:
You can use `Empty` instead of `struct{}` here if you want https://godoc.org/k8s.io/apimachinery/pkg/util/sets#Empty

Reply:
I found using `sets.Empty` a little bit confusing, so I have redefined empty locally and now it reads
`map[*node]empty`